### ISSUE TITLE
Python tests take very long

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#8494]
Looking at the test results, python tests generally take ~30 minutes to run while other wrapped languages only take a few minutes.
https://grpc-testing.appspot.com/job/gRPC_master_linux/241/testReport/(root)/aggregate_tests/

I know python has comparatively more tests (and that's great!), but it still seems that's not the only reason. Perhaps it would be worth investigating a bit why python tests take such long time to run.

(this is part of a general effort to decrease PR turnaround time)

**Comment by jtattermusch** [comment]
CC @matt-kwong 

(and sorry for typos in the issue name)

**Comment by nathanielmanistaatgoogle** [comment]
Thanks for [the link](https://grpc-testing.appspot.com/job/gRPC_master_linux/241/testReport/(root)/aggregate_tests/) but [it doesn't offer too much insight into what's going on](https://grpc-testing.appspot.com/job/gRPC_master_linux/241/testReport/(root)/aggregate_tests/run_tests_python_linux_dbg/). How much more can you tell us about what they're doing in the thirty minutes during which they're running? They should be running in parallel; do we know that they are? Is any of that thirty minutes spent in [`build_python.sh`](https://github.com/grpc/grpc/blob/master/tools/run_tests/build_python.sh), and if so how much? What's the "long pole" that if it were to be improved would most reduce the overall time?

**Comment by soltanmm-google** [comment]
Is `ccache` still being made available to the tests? Or was that removed?

(re-enabling ccache would be easier than rewriting to build using, say, `waf` to get a parallel build going)

**Comment by soltanmm-google** [comment]
Some timing data:

```
# Build w/o ccache
./tools/run_tests/build_python.sh  300.15s user 39.94s system 97% cpu 5:49.95 total

# Build w/ ccache
./tools/run_tests/build_python.sh  30.86s user 14.90s system 83% cpu 54.879 total

# Running Python tests (single job)
./tools/run_tests/run_python.sh  112.37s user 59.98s system 69% cpu 4:06.91 total

# Tests vs. test-end lag
2:31.61 total (just tests)
./tools/run_tests/run_python.sh  110.28s user 60.12s system 76% cpu 3:42.85 total

# Approximate expected build + run time
less than 12 minutes w/o ccache, less than 6 minutes w/ ccache

# run_tests w/o ccache
./tools/run_tests/run_tests.py -l python  886.71s user 191.92s system 144% cpu 12:28.24 total

# run_tests w/ ccache
./tools/run_tests/run_tests.py -l python  444.57s user 152.70s system 235% cpu 4:13.25 total

# run_tests python, docker, in sequence
./tools/run_tests/run_tests.py --use_docker -l python  0.22s user 0.15s system 0% cpu 15:17.39 total
./tools/run_tests/run_tests.py --use_docker -l python  0.18s user 0.12s system 0% cpu 3:31.19 total

# run_tests ruby, docker, in sequence (for comparison)
./tools/run_tests/run_tests.py --use_docker -l ruby  0.38s user 0.18s system 0% cpu 12:11.64 total
./tools/run_tests/run_tests.py --use_docker -l ruby  0.19s user 0.13s system 0% cpu 2:44.24 total
```

Python doesn't appear to be unreasonably slow. I posit that either Jenkins is doing something wonky that's making Python slow, or that something else has changed. In either case, I'm questioning whether or not this is further actionable without diving into Jenkins.

**Comment by jtattermusch** [comment]
@soltanmm-google   how did you obtain the data? Running locally on your workstation? 

also, what -j value has been used for run_tests.py?    If you run locally on your workstation I think you end up with -j 12 (= core count).

**Comment by jtattermusch** [comment]
Does python have many long-lived tests? If so, tweaking the -j  value in our config could have a positive effect on the total running time.

CC @matt-kwong

**Comment by matt-kwong** [comment]
I adjusted the test parallelism values to have Linux basic tests run in about 30 mins, although this doesn't affect Python specifically. I think a large reason Python tests take longer to run than other languages is that each individual test takes at ~30 seconds to run as seen here: https://grpc-testing.appspot.com/job/gRPC_master_linux/400/testReport/(root)/python_linux_dbg/. Every other language has tests finishing in < 1 second, so perhaps the slow tests issue is related to this.

**Comment by soltanmm-google** [comment]
We do have a number of tests that can take on the order of tens of seconds to finish (on my workstation). There does also exist at least one perma-hang somewhere in `_rpc_test.py` (which is being heisenbuggy much to my chagrin), but that'd result in a test timeout instead of the tests just being slow.

**Comment by jtattermusch** [comment]
@matt-kwong   perhaps it makes sense to set -j  (a.k.a. _INNER_JOBS from run_tests_matrix.py) to a higher number specifically for python? Ideally, what we would want is to set parallelism just for running the tests (not for building), but I think increasing -j  for python will do the job just fine.

**Comment by matt-kwong** [comment]
I already have _INNER_JOBS set to 150, which is pretty good imo for balancing test speed and not killing the workers. In fact, I think there might be test flakiness involved with a high _INNER_JOBS, but this is a completely separate issue I'm investigating.

**Comment by jtattermusch** [comment]
Clearly a matter of opinion, but I'd expect 150 to be way to high (because
this might also run `make -j 150` which is a bit crazy). I'd expect ~20 to
behave nicely. But yeah, let's use data and not guesses.

**Comment by matt-kwong** [comment]
I was benchmarking testing times against CPU usage on VMs, but I think that alone wasn't a good way to determine the -j value. I went ahead and readjusted it on Jenkins and will look into assignining specific -j values for different sets of tests.

**Comment by matt-kwong** [comment]
It looks like Python test timing has gone down since this issue was reported: https://grpc-testing.appspot.com/job/gRPC_master_linux/1786/testReport/(root)/. Additionally, that 10 minute report time is just the summation of the individual test times, and since we run tests in parallel, the real time taken to run all Linux Python tests is around 5 minutes. 

Running `time python tools/run_tests/run_tests.py --use_docker -t -j 16 -x report_python_linux_dbg_native_sponge_log.xml --report_suite_name python_linux_dbg_native -l python -c dbg --iomgr_platform native` takes 2.5 minutes when I run on a Linux VM. Of those 2.5 minutes, 1 minute is spent running https://github.com/grpc/grpc/blob/4f0752318b1f2f703b71b60f4404ef2954692354/tools/run_tests/helper_scripts/build_python.sh twice (once for Python 2.7 and the other for 3.4). Aside from making sure all of the commands in that script are necessary, I'm not sure how else to reduce the time taken to run Python tests.

**Comment by jtattermusch** [comment]
The right report to look at is the "aggregate tests" (each of them corresponds to a single suite ran by run_tests_matrix.py).  
https://grpc-testing.appspot.com/job/gRPC_master_linux/1786/testReport/(root)/aggregate_tests/

The sum of all tests run times is unimportant (from the high level perspective, we only care about the entire test suite finishing as soon as possible to reduce the PR turnaround time).

Actually right now  the dbg and opt  python suites take <3 mins each, which is definitely acceptable (it used to be more like 30 mins AFAIK). Do we know what we changed since then?

**Comment by jtattermusch** [comment]
I think it's fine to close this issue now, but understanding why the python tests now run 10x faster would be great.

**Comment by matt-kwong** [comment]
I'm not sure what changed, but @nathanielmanistaatgoogle might know why Python test timing has significantly decreased.
